File: grooves-api/src/main/groovy/com/github/rahulsom/grooves/api/events/RevertEvent.java
package com.github.rahulsom.grooves.api.events;

import com.github.rahulsom.grooves.api.AggregateType;

/**
 * Revert a prior event. Aggregate reverted event's effects are not applied.
 *
 * @param <AggregateT> Aggregate this event applies to
 * @param <EventIdT>   The Type for Event's {@link #getId} field
 * @param <EventT>     Event Type that could be reverted
 *
 * @author Rahul Somasunderam
 */
public interface RevertEvent<AggregateT extends AggregateType, EventIdT, EventT>
        extends BaseEvent<AggregateT, EventIdT, EventT> {

    EventIdT getRevertedEventId();
}

File: grooves-types/src/main/kotlin/com/github/rahulsom/grooves/api/events/RevertEvent.kt
package com.github.rahulsom.grooves.api.events

import com.github.rahulsom.grooves.api.AggregateType

/**
 * Revert a prior event. A reverted event's effects are not applied when computing a snapshot.
 *
 * @param [AggregateT] Aggregate this event applies to
 * @param [EventIdT] The type for [BaseEvent.id]
 * @param [EventT] The base class for all events in the hierarchy for [AggregateT].
 *
 * @author Rahul Somasunderam
 */
interface RevertEvent<
        AggregateT : AggregateType,
        EventIdT,
        EventT> :
        BaseEvent<AggregateT, EventIdT, EventT> {

    /**
     * [BaseEvent.id] from the event that should be reverted.
     */
    val revertedEventId: EventIdT
}
